File: scripts/generate_workflow.mjs
import fs from "fs";
import OpenAI from "openai";

function pickEnv(keys) {
  for (const k of keys) {
    const v = process.env[k];
    if (v && String(v).trim() !== "") return String(v).trim();
  }
  return "";
}

function extractFirstJsonObject(text) {
  // 余計な前後テキストが混ざっても、最初の { から最後の } を抜く
  const s = text.indexOf("{");
  const e = text.lastIndexOf("}");
  if (s === -1 || e === -1 || e <= s) {
    throw new Error("Model output does not contain a JSON object");
  }
  return text.slice(s, e + 1);
}

const provider = (pickEnv(["MODEL_PROVIDER"]) || "deepseek").toLowerCase();

// DeepSeek用：DEEPSEEK_API_KEY を最優先
const apiKey =
  provider === "deepseek"
    ? pickEnv(["DEEPSEEK_API_KEY"])
    : pickEnv(["OPENAI_API_KEY", "DEEPSEEK_API_KEY"]);

if (!apiKey) {
  console.error("Missing API key. Set DEEPSEEK_API_KEY (recommended) or OPENAI_API_KEY.");
  process.exit(1);
}

// DeepSeekは OpenAI互換の baseURL を https://api.deepseek.com にする
const baseURL =
  provider === "deepseek"
    ? pickEnv(["OPENAI_BASE_URL", "DEEPSEEK_BASE_URL"]) || "https://api.deepseek.com"
    : pickEnv(["OPENAI_BASE_URL"]);

const model =
  pickEnv(["MODEL_NAME", "OPENAI_MODEL"]) ||
  (provider === "deepseek" ? "deepseek-chat" : "gpt-4.1-mini");

const client = new OpenAI({
  apiKey,
  ...(baseURL ? { baseURL } : {}),
});

const n8nVersion = pickEnv(["N8N_VERSION"]) || "2.6.3";

// workflow_spec は「指示文章」(長文仕様) でOK。JSONそのものでもOK。
// ただし最終出力は“n8n import JSON”に強制する。
const spec = fs.readFileSync("workflow_spec.txt", "utf-8");

const system = `You generate a single n8n workflow IMPORT JSON compatible with n8n version ${n8nVersion}.
Hard rules:
- Output MUST be a single JSON object only. No markdown. No code fences. No commentary.
- Use only plain ASCII quotes (").
- Do NOT include smart quotes or ellipsis.
- Do NOT include unknown properties like "option". Use "options" only if that node supports it.
- Every node must have: id, name, type, typeVersion, position [x,y], parameters.
- Top-level must include: name, nodes, connections.
- Ensure it is directly importable into n8n without "Could not find property option" errors.`;

const user = `WORKFLOW_SPEC:
${spec}
`;

async function main() {
  let text = "";

  if (provider === "deepseek" || (baseURL || "").includes("deepseek")) {
    // DeepSeek: /v1/chat/completions を使う
    const res = await client.chat.completions.create({
      model,
      messages: [
        { role: "system", content: system },
        { role: "user", content: user },
      ],
      temperature: 0.2,
      max_tokens: 6000,
    });

    text = res?.choices?.[0]?.message?.content || "";
  } else {
    // OpenAI: Responses API
    const res = await client.responses.create({
      model,
      input: [
        { role: "system", content: system },
        { role: "user", content: user },
      ],
      temperature: 0.2,
      max_output_tokens: 6000,
    });

    // SDKの返り形式差異に耐える
    text =
      res?.output_text ||
      res?.output?.[0]?.content?.map((c) => c?.text).filter(Boolean).join("") ||
      "";
  }

  if (!text) throw new Error("Empty model output");

  // JSON抽出→parseで確実に「JSONだけ」にする
  const jsonStr = extractFirstJsonObject(text);

  let obj;
  try {
    obj = JSON.parse(jsonStr);
  } catch (e) {
    console.error("Model output was not valid JSON after extraction.");
    console.error("First 400 chars:", jsonStr.slice(0, 400));
    throw e;
  }

  // ここで整形して “純JSON” として保存（変な文字混入を減らす）
  fs.writeFileSync("workflow.json", JSON.stringify(obj, null, 2) + "\n", "utf-8");
  console.log("workflow.json generated.");
}

main().catch((e) => {
  console.error("FAILED:", e?.message || e);
  process.exit(1);
});

File: scripts/generate_workflow.py
import os, json, sys, textwrap, subprocess
from typing import Tuple

# Requirements:
# - env: OPENAI_API_KEY
# - optional env: OPENAI_BASE_URL (default https://api.openai.com/v1)
# - optional env: OPENAI_MODEL (default gpt-4.1-mini or gpt-4.1)
#
# This script:
# 1) Asks model to output STRICT JSON only (no markdown)
# 2) Writes workflows/workflow.json
# 3) Runs n8n import validation
# 4) If fail, feeds error back and retries up to MAX_TRIES

MAX_TRIES = int(os.getenv("MAX_TRIES", "3"))
OUT_PATH = os.getenv("OUT_PATH", "workflows/workflow.json")

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY", "")
OPENAI_BASE_URL = os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1")
OPENAI_MODEL = os.getenv("OPENAI_MODEL", "gpt-4.1")

if not OPENAI_API_KEY:
    print("Missing OPENAI_API_KEY", file=sys.stderr)
    sys.exit(2)

SPEC = os.getenv("WORKFLOW_SPEC", "").strip()
if not SPEC:
    print("Missing WORKFLOW_SPEC env", file=sys.stderr)
    sys.exit(2)

SYSTEM = (
    "You generate n8n importable workflow JSON. Output MUST be valid JSON only. "
    "No markdown. No comments. Use ASCII quotes only. No ellipsis char. "
    "Target n8n version 2.6.3. "
    "Avoid unknown properties: do not use 'option' if node expects 'options'. "
    "Use correct node typeVersion and parameter schema."
)

def call_openai(messages) -> str:
    # Uses curl to avoid adding dependencies.
    payload = {
        "model": OPENAI_MODEL,
        "messages": messages,
        "temperature": 0.2,
    }
    p = subprocess.run(
        ["curl", "-sS", f"{OPENAI_BASE_URL}/chat/completions",
         "-H", f"Authorization: Bearer {OPENAI_API_KEY}",
         "-H", "Content-Type: application/json",
         "-d", json.dumps(payload)],
        capture_output=True, text=True
    )
    if p.returncode != 0:
        print(p.stderr, file=sys.stderr)
        sys.exit(3)
    data = json.loads(p.stdout)
    return data["choices"][0]["message"]["content"]

def validate_with_n8n(path: str) -> Tuple[bool, str]:
    # Validate by importing into a temp n8n user folder
    env = os.environ.copy()
    env["N8N_USER_FOLDER"] = os.path.abspath(".n8n_tmp")
    env["N8N_ENCRYPTION_KEY"] = env.get("N8N_ENCRYPTION_KEY", "test-encryption-key-32chars!!")
    # n8n import will exit non-zero if schema invalid
    p = subprocess.run(
        ["npx", "--yes", "n8n", "import:workflow", "--input", path],
        capture_output=True, text=True, env=env
    )
    ok = (p.returncode == 0)
    out = (p.stdout or "") + "\n" + (p.stderr or "")
    return ok, out.strip()

def main():
    last_error = ""
    for i in range(1, MAX_TRIES + 1):
        user_prompt = SPEC
        if last_error:
            user_prompt += "\n\nFAILED IMPORT ERROR (fix and output corrected JSON):\n" + last_error

        messages = [
            {"role": "system", "content": SYSTEM},
            {"role": "user", "content": user_prompt},
        ]

        content = call_openai(messages).strip()

        # Hard guard: must be JSON object
        try:
            obj = json.loads(content)
        except Exception as e:
            last_error = f"Model output was not valid JSON. json.loads error: {e}"
            print(f"[try {i}] invalid JSON from model", file=sys.stderr)
            continue

        os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
        with open(OUT_PATH, "w", encoding="utf-8") as f:
            json.dump(obj, f, ensure_ascii=False, indent=2)

        ok, log = validate_with_n8n(OUT_PATH)
        if ok:
            print(f"OK: validated and wrote {OUT_PATH}")
            return  # success

        last_error = log
        print(f"[try {i}] n8n import failed:\n{log}\n", file=sys.stderr)

    print("FAILED: could not produce a valid importable workflow JSON", file=sys.stderr)
    print("Last error:\n" + last_error, file=sys.stderr)
    sys.exit(1)

if __name__ == "__main__":
    main()
